[SPARK-27070] Fix performance bug in DefaultPartitionCoalescer

fitermay opened the pull request:
When trying to coalesce a UnionRDD of two large FileScanRDDs
(each with a few million partitions) into around 8k partitions
the driver can stall for over an hour.

Profiler shows that over 90% of the time is spent in TimSort
which is invoked by `pickBin`.  This patch replaces sorting with a more
efficient `min` for the purpose of finding the least occupied
PartitionGroup

fitermay committed:
SPARK-27070: Don't use sort to find minimum PartitionGroup

When trying to coalesce a UnionRDD of two large FileScanRDDs
(each with a few million partitions) into around 8k partitions
the driver can stall for over an hour.

 Profiler shows that over 90% of the time is spent in TimSort
 which is invoked by `pickBin`.  This patch replaces sorting with a more
 efficient `min` for the purpose of finding the least occupied
 PartitionGroup

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -156,9 +156,11 @@ private[spark] class CoalescedRDD[T: ClassTag](
+
+  implicit val partitionGroupOrdering: Ordering[PartitionGroup]
+  = (o1: PartitionGroup, o2: PartitionGroup)
+  => o1.numPartitions - o2.numPartitions
```
The indent is off here. Returning the result of a subtraction in compare can overflow, though I don't think it can happen in practice here. Still, see below, I think we can just remove this.

srowen commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -219,9 +221,8 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
-    groupHash.get(key).map(_.sortWith(compare).head)
-  }
+  def getLeastGroupHash(key: String): Option[PartitionGroup] =
+    groupHash.get(key).flatMap (group => if (group.isEmpty) None else Some(group.min))
```
`.map()` should still be OK here as `None` will results in `None`. Rather than the implicit ordering, just use `.minBy`?
`.map(_.minBy(_.numPartitions))`?

srowen replied:
I'm suggesting you also not make an Option inside the function. You're creating None/Some just to flatten it again.

srowen commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -297,9 +298,8 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
-    val prefPart = if (pref == Nil) None else pref.head
-
+    val pref = currPrefLocs(p, prev).flatMap(getLeastGroupHash)
+    val prefPart = if (pref.isEmpty) None else Some(pref.min)
```
Likewise, `pref.minBy(_.numPartitions)`?

fitermay commented:
@srowen Hi, Thanks for the prompt review. I've amended the code to address your comments.

SparkQA commented:
**[Test build #4597 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4597/testReport)** for PR 23986 at commit [`4ffb772`](https://github.com/apache/spark/commit/4ffb772e19e7d1db60b9e216054931ae70b8623a).

SparkQA commented:
**[Test build #4597 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4597/testReport)** for PR 23986 at commit [`4ffb772`](https://github.com/apache/spark/commit/4ffb772e19e7d1db60b9e216054931ae70b8623a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

attilapiros commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -213,15 +210,16 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
-    groupHash.get(key).map(_.sortWith(compare).head)
-  }
+  def getLeastGroupHash(key: String): Option[PartitionGroup] =
+    groupHash
```
What about avoiding `if` and `flatmap` and `Some()` , like:
```
    groupHash.get(key).filter(_.nonEmpty).map(_.minBy(_.numPartitions))
```

srowen replied:
That seems fine to me too

attilapiros commented:
@fitermay It would be very nice to see some some rough numbers regarding the improvement caused by this PR. Could you please share us how it behaved before and after for your experiment mentioned in the description?

srowen commented:
There's a little detail in the JIRA that's pretty suggestive that this is the bottleneck; if there's a stack trace or more numbers to show, that's great. Regardless I think this a clean 'win', just a question of how big.

fitermay commented:
@srowen @attilapiros 
After digging into this further I found out what exactly is happening and why sorting here causes a major issue.

It runs out EMRFS returns the string '*' as the host of each block.  This ends up invoking the worst case of this algorithm where it tries to jam everything into the same preferred partition. In turn, this ends up running sort on hundreds thousands of records each iteration to find the minimum.  I've contacted the EMR team to suggest changing the host to 'localhost' but apparently that would break MR performance on Yarn. 

I still think this patch is a win because:

1) It's actually simpler and less code than the pre-patch code
2) There are lots of EMR users who would benefit from this until a strategic solution is found
3) It improves performance in less extreme cases as well

I will try to make the suggested changes and also generate some performance numbers for the extreme case tonight.

Thanks!

fitermay commented:
Benchmark with 100K blocks instead of several million.  Number of hosts = 1 is clearly the worst case 
```Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
Coalesced RDD:                            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Java HotSpot(TM) 64-Bit Server VM 1.8.0_112-b15 on Windows 10 10.0
Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
Coalesced RDD:                            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Coalesce Num Partitions: 100 Num Hosts: 1            492            520          33          0.2        4919.9       1.0X
Coalesce Num Partitions: 100 Num Hosts: 5            310            328          22          0.3        3103.2       1.6X
Coalesce Num Partitions: 100 Num Hosts: 10            247            267          19          0.4        2468.2       2.0X
Coalesce Num Partitions: 100 Num Hosts: 20            240            252          15          0.4        2399.7       2.1X
Coalesce Num Partitions: 100 Num Hosts: 40            229            244          16          0.4        2290.8       2.1X
Coalesce Num Partitions: 100 Num Hosts: 80            212            225          13          0.5        2123.6       2.3X
Coalesce Num Partitions: 500 Num Hosts: 1           1149           1177          26          0.1       11492.7       0.4X
Coalesce Num Partitions: 500 Num Hosts: 5            464            500          34          0.2        4643.8       1.1X
Coalesce Num Partitions: 500 Num Hosts: 10            386            397          19          0.3        3862.2       1.3X
Coalesce Num Partitions: 500 Num Hosts: 20            336            340           7          0.3        3358.1       1.5X
Coalesce Num Partitions: 500 Num Hosts: 40            269            283          17          0.4        2686.0       1.8X
Coalesce Num Partitions: 500 Num Hosts: 80            239            245           9          0.4        2391.0       2.1X
Coalesce Num Partitions: 1000 Num Hosts: 1           2213           2258          39          0.0       22131.2       0.2X
Coalesce Num Partitions: 1000 Num Hosts: 5            645            650           9          0.2        6448.8       0.8X
Coalesce Num Partitions: 1000 Num Hosts: 10            467            473           7          0.2        4673.8       1.1X
Coalesce Num Partitions: 1000 Num Hosts: 20            413            425          17          0.2        4133.7       1.2X
Coalesce Num Partitions: 1000 Num Hosts: 40            341            347          10          0.3        3412.4       1.4X
Coalesce Num Partitions: 1000 Num Hosts: 80            269            276          11          0.4        2688.8       1.8X
Coalesce Num Partitions: 5000 Num Hosts: 1          11048          11100          46          0.0      110484.2       0.0X
Coalesce Num Partitions: 5000 Num Hosts: 5           2396           2457          55          0.0       23959.0       0.2X
Coalesce Num Partitions: 5000 Num Hosts: 10           1390           1397           9          0.1       13899.1       0.4X
Coalesce Num Partitions: 5000 Num Hosts: 20            852            858           6          0.1        8516.9       0.6X
Coalesce Num Partitions: 5000 Num Hosts: 40            569            586          21          0.2        5692.7       0.9X
Coalesce Num Partitions: 5000 Num Hosts: 80            432            440           9          0.2        4322.7       1.1X
Coalesce Num Partitions: 10000 Num Hosts: 1          19685          19779          83          0.0      196853.8       0.0X
Coalesce Num Partitions: 10000 Num Hosts: 5           4044           4144          87          0.0       40437.9       0.1X
Coalesce Num Partitions: 10000 Num Hosts: 10           2393           2483          88          0.0       23931.6       0.2X
Coalesce Num Partitions: 10000 Num Hosts: 20           1242           1338          84          0.1       12419.6       0.4X
Coalesce Num Partitions: 10000 Num Hosts: 40            816            821           9          0.1        8158.7       0.6X
Coalesce Num Partitions: 10000 Num Hosts: 80            555            571          23          0.2        5554.2       0.9X
```

After patch:
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_112-b15 on Windows 10 10.0
Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
Coalesced RDD:                            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Coalesce Num Partitions: 100 Num Hosts: 1            394            433          37          0.3        3941.6       1.0X
Coalesce Num Partitions: 100 Num Hosts: 5            275            279           7          0.4        2748.4       1.4X
Coalesce Num Partitions: 100 Num Hosts: 10            236            241           9          0.4        2355.8       1.7X
Coalesce Num Partitions: 100 Num Hosts: 20            226            239          12          0.4        2259.1       1.7X
Coalesce Num Partitions: 100 Num Hosts: 40            220            233          14          0.5        2199.3       1.8X
Coalesce Num Partitions: 100 Num Hosts: 80            212            227          14          0.5        2120.3       1.9X
Coalesce Num Partitions: 500 Num Hosts: 1            961            976          24          0.1        9606.9       0.4X
Coalesce Num Partitions: 500 Num Hosts: 5            358            367          10          0.3        3580.5       1.1X
Coalesce Num Partitions: 500 Num Hosts: 10            288            299          19          0.3        2877.5       1.4X
Coalesce Num Partitions: 500 Num Hosts: 20            251            257           9          0.4        2508.5       1.6X
Coalesce Num Partitions: 500 Num Hosts: 40            248            252           4          0.4        2478.1       1.6X
Coalesce Num Partitions: 500 Num Hosts: 80            225            234          13          0.4        2247.3       1.8X
Coalesce Num Partitions: 1000 Num Hosts: 1           1575           1581           9          0.1       15747.8       0.3X
Coalesce Num Partitions: 1000 Num Hosts: 5            515            524          10          0.2        5154.8       0.8X
Coalesce Num Partitions: 1000 Num Hosts: 10            363            384          20          0.3        3633.5       1.1X
Coalesce Num Partitions: 1000 Num Hosts: 20            294            300           6          0.3        2943.6       1.3X
Coalesce Num Partitions: 1000 Num Hosts: 40            255            259           4          0.4        2549.3       1.5X
Coalesce Num Partitions: 1000 Num Hosts: 80            240            252          11          0.4        2398.7       1.6X
Coalesce Num Partitions: 5000 Num Hosts: 1           6904           6948          64          0.0       69038.0       0.1X
Coalesce Num Partitions: 5000 Num Hosts: 5           2070           2109          33          0.0       20702.0       0.2X
Coalesce Num Partitions: 5000 Num Hosts: 10           1136           1153          27          0.1       11362.4       0.3X
Coalesce Num Partitions: 5000 Num Hosts: 20            696            752          49          0.1        6964.3       0.6X
Coalesce Num Partitions: 5000 Num Hosts: 40            456            483          39          0.2        4555.8       0.9X
Coalesce Num Partitions: 5000 Num Hosts: 80            334            353          17          0.3        3340.8       1.2X
Coalesce Num Partitions: 10000 Num Hosts: 1          12789          12875         123          0.0      127889.3       0.0X
Coalesce Num Partitions: 10000 Num Hosts: 5           4040           4117          67          0.0       40402.9       0.1X
Coalesce Num Partitions: 10000 Num Hosts: 10           2141           2185          61          0.0       21414.0       0.2X
Coalesce Num Partitions: 10000 Num Hosts: 20           1152           1153           2          0.1       11516.1       0.3X
Coalesce Num Partitions: 10000 Num Hosts: 40            687            695          10          0.1        6869.5       0.6X
Coalesce Num Partitions: 10000 Num Hosts: 80            451            458           7          0.2        4505.0       0.9X
```

fitermay committed:
add benchmark and incorporate suggestion from review

fitermay commented:
@fitermay @attilapiros 
I've pushed the benchmark and the suggestion of using filter, map

attilapiros reviewed (approved):
Impressing results. 

Thanks for the improvement!

fitermay commented:
By the way. This are the results from the original PR before replacing `min` with `minBy`. It seems to be twice as fast. I'm guessing it's because of the reduction of indirection when passing an implicit ordering instead of a minBy lambda.

```
Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
Coalesced RDD:                            Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Coalesce Num Partitions: 100 Num Hosts: 1            264            289          26          0.4        2644.9       1.0X
Coalesce Num Partitions: 100 Num Hosts: 5            211            220           8          0.5        2110.6       1.3X
Coalesce Num Partitions: 100 Num Hosts: 10            215            225          10          0.5        2149.9       1.2X
Coalesce Num Partitions: 100 Num Hosts: 20            200            203           6          0.5        1996.6       1.3X
Coalesce Num Partitions: 100 Num Hosts: 40            198            205          11          0.5        1983.7       1.3X
Coalesce Num Partitions: 100 Num Hosts: 80            199            203           4          0.5        1992.8       1.3X
Coalesce Num Partitions: 500 Num Hosts: 1            465            477          15          0.2        4654.2       0.6X
Coalesce Num Partitions: 500 Num Hosts: 5            271            280          11          0.4        2707.9       1.0X
Coalesce Num Partitions: 500 Num Hosts: 10            232            250          18          0.4        2320.5       1.1X
Coalesce Num Partitions: 500 Num Hosts: 20            213            222          14          0.5        2130.8       1.2X
Coalesce Num Partitions: 500 Num Hosts: 40            210            215           9          0.5        2102.9       1.3X
Coalesce Num Partitions: 500 Num Hosts: 80            206            206           0          0.5        2062.4       1.3X
Coalesce Num Partitions: 1000 Num Hosts: 1            715            716           1          0.1        7149.7       0.4X
Coalesce Num Partitions: 1000 Num Hosts: 5            310            311           1          0.3        3098.5       0.9X
Coalesce Num Partitions: 1000 Num Hosts: 10            255            266          17          0.4        2553.8       1.0X
Coalesce Num Partitions: 1000 Num Hosts: 20            230            238          12          0.4        2304.1       1.1X
Coalesce Num Partitions: 1000 Num Hosts: 40            227            242          21          0.4        2271.1       1.2X
Coalesce Num Partitions: 1000 Num Hosts: 80            211            217          10          0.5        2114.5       1.3X
Coalesce Num Partitions: 5000 Num Hosts: 1           3043           3616         634          0.0       30428.0       0.1X
Coalesce Num Partitions: 5000 Num Hosts: 5           1035           1069          52          0.1       10353.4       0.3X
Coalesce Num Partitions: 5000 Num Hosts: 10            613            617           3          0.2        6134.6       0.4X
Coalesce Num Partitions: 5000 Num Hosts: 20            408            419          11          0.2        4082.8       0.6X
Coalesce Num Partitions: 5000 Num Hosts: 40            315            340          24          0.3        3153.3       0.8X
Coalesce Num Partitions: 5000 Num Hosts: 80            258            262           5          0.4        2577.7       1.0X
Coalesce Num Partitions: 10000 Num Hosts: 1           5385           5470         124          0.0       53848.7       0.0X
Coalesce Num Partitions: 10000 Num Hosts: 5           1856           1861           7          0.1       18561.0       0.1X
Coalesce Num Partitions: 10000 Num Hosts: 10           1022           1075          48          0.1       10223.2       0.3X
Coalesce Num Partitions: 10000 Num Hosts: 20            619            626           8          0.2        6185.8       0.4X
Coalesce Num Partitions: 10000 Num Hosts: 40            417            422           5          0.2        4168.2       0.6X
Coalesce Num Partitions: 10000 Num Hosts: 80            312            316           4          0.3        3119.6       0.8X
```

srowen commented:
Hm! That's surprising. Looking at min vs minBy, it even seems like min has more indirection (calls foldLeft). The implicit still involves calling a function to compare and get num partitions in both cases. If you're pretty sure this is accurate I'm OK returning to the implicit.

attilapiros commented:
Can I check the reason why is this difference on Monday/Tuesday? I mean can we wait with the merge?  
I am interested what causing this and would like to look into that.

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/rdd/CoalescedRDDBenchmark.scala:
```diff
@@ -0,0 +1,70 @@
+  val sc = new SparkContext(
+    "local[4]",
+    "test",
+    conf.set("spark.hadoop.fs.file.impl", classOf[DebugFilesystem].getName))
```
Could you remove this `DebugFilesystem`?

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/rdd/CoalescedRDDBenchmark.scala:
```diff
@@ -0,0 +1,70 @@
+ */
+
+package org.apache.spark.rdd
+import scala.collection.immutable
```
Please run `dev/scalastyle` and fix the Scala style issue by adding a new line.
```
$ dev/scalastyle
Scalastyle checks failed at following occurrences:
[error] /Users/dongjoon/PRS/PR-23986/core/src/test/scala/org/apache/spark/rdd/CoalescedRDDBenchmark.scala:19:0: There should at least one a single empty line separating groups java and scala.
[error] Total time: 16 s, completed Mar 8, 2019 11:33:59 AM
```

fitermay replied:
Sorry, somehow IntelliJ didn't highlight that even though I had spark scalastyle set.

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/rdd/CoalescedRDDBenchmark.scala:
```diff
@@ -0,0 +1,70 @@
+import org.apache.spark.{DebugFilesystem, SparkConf, SparkContext}
+import org.apache.spark.benchmark.{Benchmark, BenchmarkBase}
+
+object CoalescedRDDBenchmark extends BenchmarkBase {
```
Please check `KryoBenchmark` and add a benchmark description here.
- https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/serializer/KryoBenchmark.scala#L29-L40

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/rdd/CoalescedRDDBenchmark.scala:
```diff
@@ -0,0 +1,70 @@
+    val numIters = 3
+    runBenchmark("Coalesced RDD , large scale") {
+      coalescedRDD(numIters)
+    }
```
Also, we need include the generated test result like `KryoBenchmark`. If you run as a guide, you will get the generate file. Please simply add it.

dongjoon-hyun reviewed (changes requested):
Hi, @fitermay . Thank you for your first contribution. I saw the above good previous comments, and I left a few comments, too.

After fixing that, we can trigger Jenkins for your PR. Otherwise, it will fail to build.

dongjoon-hyun commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8666/
Test PASSed.

SparkQA commented:
**[Test build #103221 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103221/testReport)** for PR 23986 at commit [`0803c63`](https://github.com/apache/spark/commit/0803c63fa3178108794f24cf7e5d47e9a1f97a3f).

SparkQA commented:
**[Test build #103221 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103221/testReport)** for PR 23986 at commit [`0803c63`](https://github.com/apache/spark/commit/0803c63fa3178108794f24cf7e5d47e9a1f97a3f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103221/
Test FAILed.

fitermay commented:
> Hm! That's surprising. Looking at min vs minBy, it even seems like min has more indirection (calls foldLeft). The implicit still involves calling a function to compare and get num partitions in both cases. If you're pretty sure this is accurate I'm OK returning to the implicit.

@srowen 
There is some non-obvious indirection here.  Below is the relevant bytecode that ends up being generated

Sets up the lambda that's passed into minBy. Notice that the return type of the closure must be `Ljava/lang/Object`.  So it can't return a primitive int. 
```
    LINENUMBER 223 L0
    ALOAD 0
    INVOKEDYNAMIC apply()Lscala/Function1; [
      // handle kind 0x6 : INVOKESTATIC
      java/lang/invoke/LambdaMetafactory.altMetafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)Ljava/lang/invoke/CallSite;
      // arguments:
      (Ljava/lang/Object;)Ljava/lang/Object;, 
      // handle kind 0x6 : INVOKESTATIC
      org/apache/spark/rdd/DefaultPartitionCoalescer.$anonfun$getLeastGroupHash$3$adapted(Lorg/apache/spark/rdd/PartitionGroup;)Ljava/lang/Object;, 
      (Lorg/apache/spark/rdd/PartitionGroup;)Ljava/lang/Object;, 
      7, 
      1, 
      scala.Serializable.class, 
      1, 
      (Lorg/apache/spark/rdd/PartitionGroup;)Ljava/lang/Object;
    ]
    GETSTATIC scala/math/Ordering$Int$.MODULE$ : Lscala/math/Ordering$Int$;
    INVOKEVIRTUAL scala/collection/mutable/ArrayBuffer.minBy (Lscala/Function1;Lscala/math/Ordering;)Ljava/lang/Object;
```

The lambda first invokes the below function, whose only job is to box the primitive int
```
  // access flags 0x1019
  public final static synthetic $anonfun$getLeastGroupHash$3$adapted(Lorg/apache/spark/rdd/PartitionGroup;)Ljava/lang/Object;
    // parameter final  x$7
   L0
    LINENUMBER 223 L0
    ALOAD 0
    INVOKESTATIC org/apache/spark/rdd/DefaultPartitionCoalescer.$anonfun$getLeastGroupHash$3 (Lorg/apache/spark/rdd/PartitionGroup;)I
    INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
    ARETURN
   L1
    LOCALVARIABLE x$7 Lorg/apache/spark/rdd/PartitionGroup; L0 L1 0
    MAXSTACK = 1
    MAXLOCALS = 1
```

Then the actual method that returns `numParititons` for the comparison gets invoked.
```
 public final static synthetic $anonfun$getLeastGroupHash$3(Lorg/apache/spark/rdd/PartitionGroup;)I
    // parameter final  x$7
   L0
    LINENUMBER 223 L0
    ALOAD 0
    INVOKEVIRTUAL org/apache/spark/rdd/PartitionGroup.numPartitions ()I
    IRETURN
   L1
    LOCALVARIABLE x$7 Lorg/apache/spark/rdd/PartitionGroup; L0 L1 0
    MAXSTACK = 1
    MAXLOCALS = 1
```

fitermay committed:
SPARK-11316: Changes after review

- Use `min` with Ordering instead of `minBy` to avoid boxing overhead
- Add benchmark results
- Add benchmarks description
- Don't use DebugFilesystem in benchmark
- Fix scalastyle
- Fix some minor existing codestyle issues in CoalescedRDD.scala

SparkQA commented:
**[Test build #103271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103271/testReport)** for PR 23986 at commit [`c8424af`](https://github.com/apache/spark/commit/c8424af002d870da54f767a25c62a4579daeef2c).

fitermay commented:
@dongjoon-hyun @srowen 
I've experimented enough today to be very confident that the reason behind the performance difference is the overhead introduced by boxing/unboxing

Pushed these changes:

- Use `min` with Ordering instead of `minBy` to avoid boxing overhead
- Added benchmark results
- Added benchmarks description
- Get rid of DebugFilesystem in benchmark
- Fixed scalastyle
- Fixed some minor existing codestyle issues in CoalescedRDD.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8717/
Test PASSed.

SparkQA commented:
**[Test build #103271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103271/testReport)** for PR 23986 at commit [`c8424af`](https://github.com/apache/spark/commit/c8424af002d870da54f767a25c62a4579daeef2c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103271/
Test FAILed.

fitermay committed:
SPARK-11316: Revert change to foreach loop

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8721/
Test PASSed.

SparkQA commented:
**[Test build #103275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103275/testReport)** for PR 23986 at commit [`2566639`](https://github.com/apache/spark/commit/2566639e81a227c3161cdd02a56ac5f57065dd43).

SparkQA commented:
**[Test build #103275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103275/testReport)** for PR 23986 at commit [`2566639`](https://github.com/apache/spark/commit/2566639e81a227c3161cdd02a56ac5f57065dd43).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103275/
Test FAILed.

kiszk commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8724/
Test PASSed.

SparkQA commented:
**[Test build #103278 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103278/testReport)** for PR 23986 at commit [`2566639`](https://github.com/apache/spark/commit/2566639e81a227c3161cdd02a56ac5f57065dd43).

kiszk commented:
sounds good to me

SparkQA commented:
**[Test build #103278 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103278/testReport)** for PR 23986 at commit [`2566639`](https://github.com/apache/spark/commit/2566639e81a227c3161cdd02a56ac5f57065dd43).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103278/
Test PASSed.

fitermay commented:
@dongjoon-hyun 
I've applied your suggestions. Do you think you can give this another look?

srowen commented:
Merged to master

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -156,9 +156,11 @@ private[spark] class CoalescedRDD[T: ClassTag](
+  implicit val partitionGroupOrdering: Ordering[PartitionGroup] =
+    (o1: PartitionGroup, o2: PartitionGroup) =>
+      java.lang.Integer.compare(o1.numPartitions, o2.numPartitions)
+
```
Hi, All.
This seems to break `scala-2.11` build.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-2.7-ubuntu-scala-2.11/478/
```
[error] ../core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:161: type mismatch;
[error]  found   : (org.apache.spark.rdd.PartitionGroup, org.apache.spark.rdd.PartitionGroup) => Int
[error]  required: Ordering[org.apache.spark.rdd.PartitionGroup]
[error]     (o1: PartitionGroup, o2: PartitionGroup) =>
[error]          
```

fitermay replied:
Thanks. That’s unfortunate. I’ll fix it later tonight

dongjoon-hyun replied:
Then, let me revert this first to recover 2.11 build for the other PRs. Since this PR is already approved, I believe that the next PR will be easily accepted, @fitermay .

attilapiros replied:
@dongjoon-hyun @srowen:  Would it be a good idea to extend the PR builder to run a compile with scala 2.11 (without any test run)? 

I know it is an extra 10-15 minutes but for the 4 hours test run it might be worth preventing such situations on the other hand this must be very rare. What is your opinion?

dongjoon-hyun replied:
I agree with you @attilapiros . But, IIRC, there was a discussion on that issue and the decision at that time was the current cost is not high enough for that.
The committers have a responsibility to monitor their commit. And, we usually are able to do `HOTFIX` or revert in a short time.

attilapiros replied:
Ok, thanks. 
Yes it must be very rare.

srowen replied:
We're going to drop 2.11 support soonish anyway, so I think for now we accept the occasional breaks and fix after the fact rather than double the PR builders.

dongjoon-hyun commented:
This is reverted via https://github.com/apache/spark/commit/4bab69b22a50ae00b92ed6ab3b5120574dc3aa19 . 
Please make a PR after testing both Scala 2.11 and 2.12. Thanks!

dongjoon-hyun commented:
Scala-2.11 build is recovered and now testing is on the way.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-2.7-ubuntu-scala-2.11/482/

srowen commented:
@fitermay I guess it has to be more explicitly constructed as an `Ordering`?
```
[error] /home/jenkins/workspace/spark-master-test-maven-hadoop-2.7-ubuntu-scala-2.11/core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:161: type mismatch;
[error]  found   : (org.apache.spark.rdd.PartitionGroup, org.apache.spark.rdd.PartitionGroup) => Int
[error]  required: Ordering[org.apache.spark.rdd.PartitionGroup]
[error]     (o1: PartitionGroup, o2: PartitionGroup) =>
[error]                                              ^
[info] (org.apache.spark.rdd.PartitionGroup, org.apache.spark.rdd.PartitionGroup) => Int <: Ordering[org.apache.spark.rdd.PartitionGroup]?
[info] false
[error] one error found
[error] Compile failed at Mar 14, 2019 6:17:42 PM [22.876s]
```

fitermay commented:
@srowen Yes, Scala 2.11 won't do the SAM conversion to ordering. 
Sent another PR tested against both Scala versions https://github.com/apache/spark/pull/24116
